PROYECTO ETAPA 1. SOLUCIÓN AL PROBLEMA DE PREDICCIÓN DE DEMANDA DE ALQUILES DE BICICLETAS IMPLEMENTANDO
REGRESIÓN LASSO Y POLINOMIAL

Antes que nada, hablemos de los datos, pues hay que tener en cuenta varios factores antes de empezar a 
codificar.
1. Vienen en formato xlsx y se quieren implementar en csv.
2. Se tienen un total de 17379 registros y 9 variables (columnas) para alimentar nuestros modelos.
3. La variable objetivo se denota como "cnt".
4. Hay la presencia de variables categóricas como "season", "weathersit" y "time_of_day" las cuales se
tienen que transformar a numéricas.
5. Dichas variables categóricas son nominales, ya que no cuentan con un orden jerárquico que las distinga
esto quiere decir que su transformación se puede hacer más directa sin la preocupación de mantener un orden.
6. Hay que evaluar el tipo de transformación que se aplicará a las variables categóricas buscando la mejor
opción.

1. Se realizan las importaciones necesarias para la manipulación de los datos y construcción de ambos modelos a implementar

In [30]:
import pandas as pd

from sklearn.model_selection import train_test_split, KFold, GridSearchCV
from sklearn.preprocessing import StandardScaler, PolynomialFeatures, RobustScaler
from sklearn.linear_model import Lasso
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, r2_score, root_mean_squared_error

2. Para este caso en específico se habían proporcionado los datos en formato xlsx por lo cual se convirtió a un csv.
Sin embargo, al transformarlo el archivo contenía "" lo que impedía la lectura apropiada del archivo
por lo cual se procede desde código a acceder al archivo y realizar el ajuste eliminando dichas ""

In [31]:
path = '../resources/Datos_Etapa_1_csv.csv'
with open(path, 'r') as file:
    content = file.read()

content = content.replace('"', '')  # Eliminar las comillas dobles

with open(path, 'w') as file:
    file.write(content)

3. Se realiza la carga de los datos y mostramos la información para verificar que la información esté correcta

In [32]:
#Carga y verificación del CSV 
data_original = pd.read_csv(path, sep=',')
data_copy = data_original.copy() #---- variable para manejo de transformación get_dummies
data_copy2 = data_original.copy() #--- variable para manejo de transformación .cat.codes
data_copy.head()

season  weekday weathersit  temp   atemp   hum  windspeed  cnt time_of_day
0  Winter        6      Clear  3.28  3.0014  0.81        0.0   16       Night
1  Winter        6      Clear  2.34  1.9982  0.80        0.0   40       Night
2  Winter        6      Clear  2.34  1.9982  0.80        0.0   32       Night
3  Winter        6      Clear  3.28  3.0014  0.75        0.0   13       Night
4  Winter        6      Clear  3.28  3.0014  0.75        0.0    1       Night

4. para este paso se plantearon dos diferentes transformaciones, la primera un pd.get_dummies() y segundo un .cat.codes
(para probar cada transformación descomentar y comentar las líneas correspondientes a la transformation a probar)
la diferencia entre estos dos es que getdummies genera columnas binarias adicionales según el número de variables y el número
de asignaciones posibles, esto podría significar un aumento en los tiempos de respuesta al tener que procesar más información.
Por otro lado, implementar .cat.codes genera una transformación más cruda, es decir asigna valores directamente a las clasificaciones
por ejemplo los días de la semana (lunes a viernes) los asigna como 1 a 5, esto teóricamente no sería lo correcto,
dado que la máquina lo interpreta como si un día valiera más que otro y no como identificadores. Sin embargo, con objetivo de pruebas
y tener un resultado adicional para comparar se ha aplicado a los modelos.


In [33]:
print(data_copy.shape)
#----------------------------------------------------------------------------------transformación .cat.codes
#data_copy2['season'] = data_copy['season'].astype('category').cat.codes
#data_copy2['weathersit'] = data_copy['weathersit'].astype('category').cat.codes
#data_copy2['time_of_day'] = data_copy['time_of_day'].astype('category').cat.codes
#encoded_data = data_copy2
#---------------------------------------------------------------------------------Transformación get_dummies
categorical_columns = ['weekday','season', 'weathersit', 'time_of_day']
encoded_data = pd.get_dummies(data_copy, columns=categorical_columns)
categorical_encoded_columns = [col for col in encoded_data.columns if col not in data_copy.columns.difference(categorical_columns)]
encoded_data[categorical_encoded_columns] = encoded_data[categorical_encoded_columns].astype(int)   
#--------------------------------------------------------------------------------------------------------------------------
data_copy = encoded_data
data_copy.head()    

(17379, 9)


temp   atemp   hum  windspeed  cnt  weekday_0  weekday_1  weekday_2  \
0  3.28  3.0014  0.81        0.0   16          0          0          0   
1  2.34  1.9982  0.80        0.0   40          0          0          0   
2  2.34  1.9982  0.80        0.0   32          0          0          0   
3  3.28  3.0014  0.75        0.0   13          0          0          0   
4  3.28  3.0014  0.75        0.0    1          0          0          0   

   weekday_3  weekday_4  ...  season_Spring  season_Summer  season_Winter  \
0          0          0  ...              0              0              1   
1          0          0  ...              0              0              1   
2          0          0  ...              0              0              1   
3          0          0  ...              0              0              1   
4          0          0  ...              0              0              1   

   weathersit_Clear  weathersit_Heavy Rain  weathersit_Light Rain  \
0                 1                      0                      0   
1                 1                      0                      0   
2                 1                      0                      0   
3                 1                      0                      0   
4                 1                      0                      0   

   weathersit_Mist  time_of_day_Evening  time_of_day_Morning  \
0                0                    0                    0   
1                0                    0                    0   
2                0                    0                    0   
3                0                    0                    0   
4                0                    0                    0   

   time_of_day_Night  
0                  1  
1                  1  
2                  1  
3                  1  
4                  1  

[5 rows x 23 columns]

5. Se hace la verificación en los registros para que no hayan datos vacíos, en este caso no se tienen datos vacíos pór lo cual no requiere
hacer ningún ajuste

In [34]:
data_copy.isna().sum()

temp                     0
atemp                    0
hum                      0
windspeed                0
cnt                      0
weekday_0                0
weekday_1                0
weekday_2                0
weekday_3                0
weekday_4                0
weekday_5                0
weekday_6                0
season_Fall              0
season_Spring            0
season_Summer            0
season_Winter            0
weathersit_Clear         0
weathersit_Heavy Rain    0
weathersit_Light Rain    0
weathersit_Mist          0
time_of_day_Evening      0
time_of_day_Morning      0
time_of_day_Night        0
dtype: int64

6. Verificamos si existen registros duplicados, a lo cual nos damos cuenta qu existen 42.

In [35]:
#verificar registros duplicados 
data_copy.duplicated().sum()

np.int64(42)

7. Procedemos a eliminar los datos duplicados y verificar su correcto proceso mirando el número de registros

In [36]:
#eliminar registros replicados
data_copy = data_copy.drop_duplicates()
print(data_copy.shape)

(17337, 23)


# REGRESSION CON REGULARIZACIÓN LASSO

8. Construimos y entrenamos el primer modelo para este caso se implementa Lasso con las parametrizaciones solicitadas
como el random_state=77 y los valores alpha en 1, 2, 3, 4, 5, se utiliza una normalización de los datos estandar con 
StandardScaler

In [37]:
#SEPARCION DE VARIABLE OBJETIVO DE INDEPENDIENTES   ---------------------------------------------- REGRESION CON REGULACION LASSO
train_lasso, test_lasso = train_test_split(data_copy, test_size=0.2, random_state=77)
x_train_lasso = train_lasso.drop(['cnt'], axis = 1)
y_train_lasso = train_lasso['cnt']
#ESTANDARIZACIÓN DE LOS DATOS
columns = x_train_lasso.columns  
#creacion de objeto StandardScaler()
scaler = StandardScaler()
x_train_lasso = pd.DataFrame(scaler.fit_transform(x_train_lasso), columns = columns)
#BUSQUEDA DE HIPERPARAMETRO Y ENTRENAMIENTO
kfold = KFold(n_splits=10, shuffle= True, random_state=77)
#creacion objeto clase lasso
lasso = Lasso()
#definición de busqueda de hiperparametro
valores_alpha = [1, 2, 3, 4, 5]
param_grid_lasso = {'alpha': valores_alpha}
grid_lasso = GridSearchCV(lasso, param_grid_lasso, cv = kfold, n_jobs = -1, scoring= 'neg_mean_squared_error')
grid_lasso.fit(x_train_lasso, y_train_lasso)


GridSearchCV(cv=KFold(n_splits=10, random_state=77, shuffle=True),
             estimator=Lasso(), n_jobs=-1,
             param_grid={'alpha': [1, 2, 3, 4, 5]},
             scoring='neg_mean_squared_error')

9. Iniciamos las pruebas con el modelo entrenado, mostramos el mejor parámetro alpha y los coéficientes hallados
en este caso el modelo determina que el mejor valor para alpha es 1 y el valor de cada uno de los coeficientes 

In [38]:
#mostrar mejor valor alfa 
mejor_modelo_lasso = grid_lasso.best_estimator_
x_test_lasso = test_lasso.drop(['cnt'], axis = 1)
y_test_lasso = test_lasso['cnt']
x_test_lasso = pd.DataFrame(scaler.transform(x_test_lasso), columns= columns)

print("mejor parametro lasso:", grid_lasso.best_params_)
list(zip(x_train_lasso.columns, mejor_modelo_lasso.coef_))

mejor parametro lasso: {'alpha': 1}


[('temp', np.float64(39.191981038398495)),
 ('atemp', np.float64(21.842442433534956)),
 ('hum', np.float64(-27.64610970016669)),
 ('windspeed', np.float64(-0.0)),
 ('weekday_0', np.float64(-2.426884847183269)),
 ('weekday_1', np.float64(-0.5682387396568973)),
 ('weekday_2', np.float64(-0.0)),
 ('weekday_3', np.float64(0.0)),
 ('weekday_4', np.float64(0.0)),
 ('weekday_5', np.float64(0.5062339724906544)),
 ('weekday_6', np.float64(0.6723201777679966)),
 ('season_Fall', np.float64(22.548732507426287)),
 ('season_Spring', np.float64(8.41304080787491)),
 ('season_Summer', np.float64(-2.302852369371599)),
 ('season_Winter', np.float64(-2.3308152040124495)),
 ('weathersit_Clear', np.float64(1.1937844567950389)),
 ('weathersit_Heavy Rain', np.float64(0.0)),
 ('weathersit_Light Rain', np.float64(-10.59687737334932)),
 ('weathersit_Mist', np.float64(-0.0)),
 ('time_of_day_Evening', np.float64(39.64565520338253)),
 ('time_of_day_Morning', np.float64(0.0)),
 ('time_of_day_Night', np.float64(-48.6

10. Hacemos pruebas de test y valuamos cada uno de los errores 

In [39]:
#predicciones
y_pred_lasso = mejor_modelo_lasso.predict(x_test_lasso)
#manejor de errores
rmse_lasso = root_mean_squared_error(y_test_lasso, y_pred_lasso)
mae_lasso = mean_absolute_error(y_test_lasso, y_pred_lasso)
r2_lasso = r2_score(y_test_lasso, y_pred_lasso)

print(y_pred_lasso)
print(["RMSE:", rmse_lasso], ["MAE:", mae_lasso], ["R2:", r2_lasso])

[ 88.06062622 142.09409283 246.10601738 ... 245.07867842 258.47375809
 183.58233138]
['RMSE:', np.float64(139.73457240621644)] ['MAE:', np.float64(103.60265964565383)] ['R2:', 0.41915267678849677]


# REGRESIÓN POLINOMIAL

11.  Ahora se procede a construir el segundo modelo con las especificaciónes solisitadas. El random_state = 77 y los valores
de búsqueda 2,3. Adicionalmente en el pipeline se ubica primero el RobustScaler dentro del orden para normalizar primero
y luego si implementar el aumento de polinomio, esto con el fin de obtener mejores resultados, ya que se probaron ambos casos
y los resultados mejoraron en este orden.
Imprimiendo en pantalla el valor del mejor grado de polinomio podemos darnos cuenta de que dicho valor es 3

In [43]:
#SEPARCION DE VARIABLE OBJETIVO DE INDEPENDIENTES Y SEPARACIÓN TEST ENTRENAMIENTO ---------------------------------------------- POLINOMIAL
x_pol = data_copy.drop(['cnt'], axis=1)
y_pol = data_copy['cnt']
x_train_pol, x_test_pol, y_train_pol, y_test_pol = train_test_split(x_pol, y_pol, test_size=0.2, random_state=77)
#BUSQUEDA DE HIPER PARAMETRO
polynomial_regression = make_pipeline(
    RobustScaler(),
    PolynomialFeatures(),
    LinearRegression()
)
kfold = KFold(n_splits=10, shuffle= True, random_state=77)
#espacio de busqueda 
valores_busqueda = [2,3]
param_grid_pol = {'polynomialfeatures__degree': valores_busqueda}
grid_pol = GridSearchCV(polynomial_regression, param_grid_pol, cv = kfold, n_jobs = 1, scoring= 'neg_mean_squared_error')
grid_pol.fit(x_train_pol, y_train_pol)

print("mejor parametro polinomial: ", grid_pol.best_params_)

mejor parametro polinomial:  {'polynomialfeatures__degree': 3}


12. Se hacen las pruebas de test y finalmente se imprimen los resultados de los errores RMSE, MAE y R2

In [44]:
mejor_modelo_pol = grid_pol.best_estimator_
#prediciones
y_pred_pol = mejor_modelo_pol.predict(x_test_pol)
#manejo de errores
rmse_pol = root_mean_squared_error(y_test_pol, y_pred_pol)
mae_pol = mean_absolute_error(y_test_pol, y_pred_pol)
r2_pol = r2_score(y_test_pol, y_pred_pol)

print(y_pred_pol)
print(["RMSE:", rmse_pol], ["MAE",mae_pol], ["R2:", r2_pol])

[ -2.25   176.75   265.25   ... 249.     216.125  223.4375]
['RMSE:', np.float64(130.53181923152522)] ['MAE', np.float64(94.28348471741637)] ['R2:', 0.4931411506040242]


13. En esta celda se puede observar la comparativa entre los resultados de ambos modelos según la transformación aplicada a los datos

In [45]:
print("ERRORES LASSO:")
print(["RMSE:", rmse_lasso], ["MAE",mae_lasso], ["R2:", r2_lasso])
print("ERRORES POLYNOMIAL:")
print(["RMSE:", rmse_pol], ["MAE:", mae_pol], ["R2:", r2_pol])

ERRORES LASSO:
['RMSE:', np.float64(139.73457240621644)] ['MAE', np.float64(103.60265964565383)] ['R2:', 0.41915267678849677]
ERRORES POLYNOMIAL:
['RMSE:', np.float64(130.53181923152522)] ['MAE:', np.float64(94.28348471741637)] ['R2:', 0.4931411506040242]


## CONCLUSIONES

#### GET_DUMMIES
- ERRORES LASSO:
    - RMSE: 139.73457240621644 MAE: 103.60265964565383 R2: 0.41915267678849677
- ERRORES POLYNOMIAL:
    - RMSE: 130.53181923152522 MAE: 94.28348471741637 R2: 0.4931411506040242

#### CAT.CODES
- ERRORES LASSO:
    - RMSE: 140.2024018429416 MAE: 103.80299404474125 R2: 0.41525682828042443
- ERRORES POLYNOMIAL:
    - RMSE: 133.10371634273045 MAE: 96.77711637417589 R2: 0.4729708769242106

- En la parte superior se pueden observar los resultados capturados para ambos modelos implementando ambas transformaciones, dejando en evidencia que para este caso
el modelo polinomial con grado 3 y transformación de get_dummies presenta una mejoría superior respecto.
- Analizando los resultados, evidentemente los modelos no son los mejores, pues solamente se logró un r2 máximo de 0.49 que se podría interpretar como que el modelo
un poquito menos de la mitad de las veces acierta en su predicción adicionalmente sus RMSE y MAE son elevados, ya que estos deberían tender a 0 esto indica que la
desviación en de las predicciones respecto al valor real es bastante elevada.
- En términos de lasso nos podemos dar cuenta que el mejor valor alpha proporcionado por el modelo es el 1
- Aplicando el modelo lasso a los datos implementando transformación cat.codes o get_dummies pudimos darnos cuenta de que había una persistencia de 0.0 en una variable
llamada windspeed esto podría significar que dicha variable no representa un valor agregado a la funcionalidad del modelo, es decir, no aporta para que el modelo
aprenda. Por otro lado, utilizando get_dummies nos podemos dar cuenta que el número de coeficientes que se vuelven 0 es mucho mayor siendo un total de 7, no obstante
haciendo pruebas eliminándolas y por análisis si bien las variables categóricas se dividen en más columnas estas siguen teniendo una relación intrínseca entre ellas
un claro ejemplo es el weekday que tiene varias columnas (7 en total) donde cuatro de ellas tienen coeficientes en 0, al quitar dichas columnas en 0 el modelo en vez
de mejorar, empeora. La manera de poder eliminar la variable completa es que todas sus categorías estén en cero o muy cercanas.
- Entender los datos es fundamental, el impacto que tienen en el modelo, y saber qué tipo de transformación aplicar a los datos y a que variables en específico hacen
la diferencia en tener un modelo cada vez mejor entrenado, adicionalmente saber y conocer que tipo de estandarización aplicar permite que el modelo pueda interpretar
mejor los datos debido a que al cambiar entre RobustScaler, StandarScaler, etc. las variaciones de los errores se ve afectada para mejor o para peor.